source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1024,7 +1025,191 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::httpAsyncClient
+ClusterManagerImpl::ClusterDiscoveryManager::addCallback(
+    const std::string& name, const ClusterDiscoveryCallbackSharedPtr& callback) {
+  ENVOY_LOG(trace, "cm cdm: adding callback for the cluster name {} in {}", name,
+            parent_.thread_local_dispatcher_.name());
+  auto& callbacks_list = pending_clusters_[name];
+  auto it = callbacks_list.emplace(callbacks_list.end(), callback);
+  auto handle = std::make_unique<ClusterDiscoveryCallbackHandleImpl>(*this, name, it);
+  auto discovery_in_progress = (callbacks_list.size() > 1);
+  return {std::move(handle), discovery_in_progress};
+}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::erase(const std::string& name,
+                                                        CallbackListIterator it) {
+  ENVOY_LOG(trace, "cm cdm: dropping callback for the cluster name {} in ", name,
+            parent_.thread_local_dispatcher_.name());
+  const bool drop_list = eraseFromList(name, it);
+  if (drop_list) {
+    ENVOY_LOG(trace, "cm cdm: dropped last callback for the cluster name {} in {}", name,
+              parent_.thread_local_dispatcher_.name());
+    pending_clusters_.erase(name);
+  }
+}
+
+bool ClusterManagerImpl::ClusterDiscoveryManager::eraseFromList(const std::string& name,
+                                                                CallbackListIterator it) {
+  auto map_it = pending_clusters_.find(name);
+  if (map_it == pending_clusters_.end()) {
+    ENVOY_LOG(trace,
+              "cm cdm: no callbacks for the cluster name {} found, no callbacks to remove in {}",
+              name, parent_.thread_local_dispatcher_.name());
+    return false;
+  }
+  auto& list = map_it->second;
+  list.erase(it);
+  return list.empty();
+}
+
+OdCdsApiHandlePtr
+ClusterManagerImpl::allocateOdCdsApi(const envoy::config::core::v3::ConfigSource& odcds_config,
+                                     OptRef<xds::core::v3::ResourceLocator> odcds_resources_locator,
+                                     ProtobufMessage::ValidationVisitor& validation_visitor) {
+  // TODO(krnowak): Instead of creating a new handle every time, store the handles internally and
+  // return an already existing one if the config or locator matches. Note that this may need a way
+  // to clean up the unused handles, so we can close the unnecessary connections.
+  auto odcds = OdCdsApiImpl::create(odcds_config, odcds_resources_locator, *this, stats_,
+                                    validation_visitor);
+  return OdCdsApiHandleImpl::create(*this, std::move(odcds));
+}
+
+ClusterDiscoveryCallbackHandlePtr ClusterManagerImpl::requestOnDemandClusterDiscovery(
+    OdCdsApiSharedPtr odcds, const std::string& name, ClusterDiscoveryCallbackSharedPtr callback,
+    std::chrono::milliseconds timeout) {
+  ThreadLocalClusterManagerImpl& cluster_manager = *tls_;
+
+  auto [handle, discovery_in_progress] = cluster_manager.cdm_.addCallback(name, callback);
+  if (discovery_in_progress) {
```

Comment:
Might be useful to add a comment that this will only catch discoveries in progress in this thread; But if a reponse to the same request made in another thread arrives earlier, the callback will be invoked anyway.

Reply:
Will add it.